File: NamasteDSA/Arrays/MergeSortedArrays.java
public class MergeSortedArrays {
    public static void mergeSortedArrays(int[] array1, int m, int[] array2, int n) {
        int pos1 = m - 1;
        int pos2 = n - 1;

        for (int i = m + n - 1; i >= 0; i--) {
            if (pos2 < 0) {
                break;
            }

            if (pos1 >= 0 && array1[pos1] > array2[pos2]) {
                array1[i] = array1[pos1];
                pos1--;
            } else {
                array1[i] = array2[pos2];
                pos2--;
            }
        }

        for (int i = 0; i < array1.length; i++) {
            System.out.println(array1[i]);
        }
    }

    public static void main(String[] args) {
        int[] array1 = { 1, 2, 3, 0, 0, 0 };
        int[] array2 = { 2, 5, 6 };
        mergeSortedArrays(array1, 3, array2, 3);

        int[] array3 = { 1 };
        int[] array4 = {};
        mergeSortedArrays(array3, 1, array4, 0);

        int[] array5 = { 0 };
        int[] array6 = { 1 };
        mergeSortedArrays(array5, 0, array6, 1);

        int[] array7 = { 1, 2, 0, 0, 0 };
        int[] array8 = { 2, 5, 6 };
        mergeSortedArrays(array7, 2, array8, 3);
    }
}

File: NamasteDSA/Arrays/mergeSortedArrays.js
function mergeSortedArrays(array1, m, array2, n) {
  let p1 = 0;
  let p2 = 0;
  let array1Copy = array1.slice(0, m);

  for (let index = 0; index < m + n; index++) {
    if (p2 >= n || (p1 < m && array1Copy[p1] < array2[p2])) {
      array1[index] = array1Copy[p1];
      p1++;
    } else {
      array1[index] = array2[p2];
      p2++;
    }
  }

  return array1;
}

function mergeSortedArrays2(array1, m, array2, n) {
  let p1 = m - 1;
  let p2 = n - 1;

  for (let index = m + n - 1; index >= 0; index--) {
    if (p2 < 0) {
      break;
    }

    if (p1 >= 0 && array1[p1] > array2[p2]) {
      array1[index] = array1[p1];
      p1--;
    } else {
      array1[index] = array2[p2];
      p2--;
    }
  }

  return array1;
}

console.log(mergeSortedArrays2([1, 2, 3, 0, 0, 0], 3, [2, 5, 6], 3));
console.log(mergeSortedArrays2([1], 1, [], 0));
console.log(mergeSortedArrays2([0], 0, [1], 1));
console.log(mergeSortedArrays2([1, 2, 0, 0, 0], 2, [2, 5, 6], 3));
